### PR TITLE
🌱 Implement runtime-openapi-gen

### PR DESCRIPTION
**Pull request by sbueringer**
I would suggest to merge this here on the fork. This gives us already the tool to generate proposal-quality OpenAPI YAMLs for Swagger.

I intentionally added a few TODOs which are relevant for the PR against upstream CAPI. Happy to add more if necessary.

I think it's okay to add very initial topology mutation / lifecycle hooks types. We will clean them up when we tackle those parts. But we can already have them for the proposals (I think the lifecycle hooks need a follow-up PR to get to the proposal state).

For the upstream PR:

We can take this PR as a basis for a PR to get runtime-openapi-gen merged, with the following open tasks:
* [ ] the topology mutation and cluster lifecycle types should not be part of the runtime-openapi-gen PR
* We should
  * [ ] have clean error messages
  * [ ] consider logging a bit more / at all
  * [ ] add some unit tests
  * [ ] add a bit more inline documentation
  * [ ] Grep for `TODO(openapi)`, discuss open topics, decide which ones should be implemented in the first iteration vs. later

### CODE REVIEW

File: internal/runtime/catalog/openapi.go [hunk 1]
```diff
@@ -20,172 +20,228 @@ import (
...
+		},
+	}
+	path := GVHToPath(gvh, "")
+
+	// Add name parameter to operation path if necessary.
+	// TODO(openapi): TBD if we want to add the name parameter.
+	// I think it's useful as that reflects the real path which Runtime Extensions will handle.
+	// Follow-up: potentially drop Singleton field.
+	if !hookDescriptor.metadata.Singleton {
+		path = GVHToPath(gvh, "{name}")
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/runtime/catalog/openapi.go:87]: Yes

[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/runtime/catalog/openapi.go:87]: Hm it should be only the name from discovery not the concatenated one. 

The RuntimeExtension implementation should not depend on the name of the CR

**Review by killianmuldoon — COMMENTED** [review]
/lgtm

Good to merge I think. We can resolve outstanding questions as we go on.

File: exp/runtime/hooks/api/v1alpha1/topologymutation_types.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+// GeneratePatchesRequest is the request of the GeneratePatches hook.
+// +kubebuilder:object:root=true
+type GeneratePatchesRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Variables are global variables for all templates.
+	Variables []Variable `json:"variables"`
+
+	// Items is the list of templates to generate patches for.
+	Items []GeneratePatchesRequestItem `json:"items"`
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/topologymutation_types.go:37]: I think I would defer. We could set it, but we would really only set it to get rid of the warning of openapi-gen, which seems like a bad reason to do it :)

[the remark this answers is not in the document]
Comment (killianmuldoon) [reply exp/runtime/hooks/api/v1alpha1/topologymutation_types.go:37]: Sure thing - just wasn't sure if I was missing something.

File: poc/test/rte-call-webhookregistration/main.go [hunk 3]
```diff
@@ -57,10 +57,9 @@ func main() {
 			Host(r.host).
 			Build()
 
-		svc := &v1alpha32.DiscoveryHook{}
 		in := &v1alpha32.DiscoveryHookRequest{First: 1, Second: "Hello CAPI runtime extensions!"}
 		out := &v1alpha32.DiscoveryHookResponse{}
-		if err := client.ServiceOld(svc, http.SpecVersion(r.version)).Invoke(ctx, in, out); err != nil {
+		if err := client.ServiceOld(v1alpha32.Discovery, http.SpecVersion(r.version)).Invoke(ctx, in, out); err != nil {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply poc/test/rte-call-webhookregistration/main.go:62]: Absolutely, at some point we should enforce import aliases, but isn't really high prio.

File: internal/runtime/catalog/openapi.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ykakarap) [reply internal/runtime/catalog/openapi.go:87]: >The RuntimeExtension implementation should not depend on the name of the CR

Good point.

**Review by ykakarap — APPROVED** [review]
/lgtm 

This looks good for open api work. As discussed, lets do follow-up other parts that need to be adjusted (eg, Lifecycle hooks, etc).

**Pull request merged**